### ISSUE TITLE
Possible inconsistency of SA1117

### ISSUE THREAD

**Issue by Piedone** [DotNetAnalyzers/StyleCopAnalyzers#3183]
The `Invoked()` invocation in the following code will correctly cause an [SA1117](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1117.md) violation:

```C#
    public class Ok
    {
        public void Test() =>
            Invoked("first argument", input =>
            {
            }, "third argument");

        private void Invoked(string firstParameter, Action<int> action, string thirdParameter)
        {
        }
    }
```

![image](https://user-images.githubusercontent.com/1976647/89554246-e61ca480-d80e-11ea-8501-b99665fdbe9a.png)

However, the following class won't cause the same violation, though I think the body of `Test()` is wrongly formatted in the same way:

```C#
    public class NotOk
    {
        public void Test() =>
            Invoked("first argument", input =>
            {
            });

        private void Invoked(string firstParameter, Action<int> action)
        {
        }
    }
```

Could you please advise? Did I misconfigure something or is this an inconsistency in the SA1117 analyzer?

May I add that the following code causes a violation correctly as well but for SA1116:

```C#
    public class OkToo
    {
        public void Test() =>
            Invoked(input =>
            {
            }, "second argument");

        private void Invoked(Action<int> action, string secondParameter)
        {
        }
    }
```

![image](https://user-images.githubusercontent.com/1976647/89554378-12d0bc00-d80f-11ea-9d64-d6b59f132803.png)

Thank you!

**Comment by pantosha** [comment]
Try PR #3115. It should help.

**Comment by Piedone** [comment]
I think this is only partially a duplicate of https://github.com/DotNetAnalyzers/StyleCopAnalyzers/issues/3106 (there are no concatenated strings in the parameters here) but the PR fixes the issue perfectly it seems! I've tried out the `NotOk` class in a unit test and gives the warning as it should.

**Comment by sharwell** [comment]
As best I can tell, this is the expected behavior of both SA1116 and SA1117.

**Comment by Piedone** [comment]
Could you explain why is that? As I see it, in both my above examples parameters aren't placed on the same line but only one of them raises a violation.

**Comment by sharwell** [comment]
Both rules interpret "placed on" to mean "the line where it starts". In most cases, the use of a multi-line argument will force the invocation to use the "one line per argument" style. The one exception to this is a case where the only multi-line argument is the last argument.

**Comment by Piedone** [comment]
That exception explains it. Why the exception though? In any case, #3115 seems to give a violation for my second example too.

**Comment by sharwell** [comment]
> In any case, #3115 seems to give a violation for my second example too.

I'm in the process of excluding this specific case from that PR right now actually 😄

**Comment by Piedone** [comment]
I don't see why there's an exception though. "If it fits into a single line, then put all arguments into the same line, otherwise each argument into their own lines" is a simple rule to follow, and quite logical IMO. Making an exception for the last argument seems counter-intuitive and I don't see the readability advantage either.

**Comment by sharwell** [comment]
The issue here is there are _thousands_ of projects using these analyzers. It seems likely that changing the behavior would mean new diagnostics show up in a significant number of projects. I find the argument that #3106 is a bug that users expect to be fixes more compelling than an argument that a lambda as the last parameter would need to be split out.

**Comment by Piedone** [comment]
Wouldn't it be possible to add this as a configurable check for a future release, so whoever wants it can accept the breaking change?

**Comment by mathew-tolj** [comment]
For those of us who would like to use the strict definition of multiline parameters (all parameters on a single line, otherwise each parameter on a new line) would implementing new rules in the alternate rules category (i.e. SX1116, SX1117) be an acceptable solution? That way existing projects would not have any changes to worry about but could opt in to the stricter definition. I know I would opt in at my workplace as we find it a bit inconsistent that following code would not raise any warnings:

```csharp
Invoked(input =>
    {
    });
```

but the following code would:

```csharp
Invoked(input =>
    {
    },
    "first argument",
   "third argument");
```

Under the strict definition I would expect this code to look as follows:

```csharp
Invoked(
    input =>
    {
    });

Invoked(
    input =>
    {
    },
    "first argument",
   "third argument");
```

I would be happy to try and submit a pull request for those changes.
